content/blog/quarkus-and-testcontainers.asciidoc

```diff
@@ -0,0 +1,142 @@
+---
+title: "Quarkus and Testcontainers"
+date: 2020-11-28T17:45:00+01:00
+draft: false
+---
+:source-highlighter: rouge
+:rouge-style: base16.dark
+:icons: font
+:imagesdir: /images
+ifdef::env-github[]
+:imagesdir: ../../static/images
+endif::[]
+
+The https://www.testcontainers.org/[Testcontainers] project is invaluable for spinning up containerized resources during your (JUnit) tests,
+e.g. databases or Kafka clusters.
+
+The project provides the https://www.testcontainers.org/quickstart/junit_5_quickstart/[`@Testcontainers`] JUnit extension for controlling the lifecycle of containers used by a test.
+When testing a https://quarkus.io/[Quarkus] application though, this is at odds somewhat with Quarkus' own https://quarkus.io/guides/getting-started-testing#recap-of-http-based-testing-in-jvm-mode[`@QuarkusTest`] extension;
+it's a best practice to avoid fixed ports for any containers started by Testcontainers.
+Instead, you should rely on Testcontainers to automatically allocate random free ports.
```

Comment:
```suggestion
Instead, you should rely on Docker to automatically allocate random free ports.
```

Reply:
Oh, I didn't know that. So Docker itself can choose a public exposed port?

Reply:
yes, that's what happens when you do `docker run -p 6379 redis` - a random port is assigned by Docker :)

Reply:
Yeah, TIL. I always was using -p 1234:1234 :)